source/common/router/router.cc

```diff
@@ -1003,6 +1009,10 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::HeaderMapPtr&& head
 
   modify_headers_(*headers);
 
+  absl::optional<Grpc::Status::GrpcStatus> grpc_status = Grpc::Common::getGrpcStatus(*headers);
```

Comment:
We already have gRPC logic in `handleNon5xxResponseHeaders`. Can we unify this logic and only do this computation once vs. in multiple places?

Reply:
Done.
I need to move this logic out from handleNon5xxResponseHeaders() to onUpstreamHeaders() since there are some scenarios when handleNon5xxResponseHeaders() won't be called.